src/comic/wallet-comic.service.ts

```diff
@@ -132,29 +133,34 @@ export class WalletComicService {
     return walletComic;
   }
 
-  async toggleSubscribe(walletAddress: string, comicSlug: string) {
+  async toggleAction({
+    walletAddress,
+    comicSlug,
+    payload,
+  }: {
+    walletAddress: string;
+    comicSlug: string;
+    payload: Record<string, boolean | Date>;
+  }) {
     let walletComic = await this.prisma.walletComic.findUnique({
       where: { comicSlug_walletAddress: { walletAddress, comicSlug } },
     });
+    if (!walletComic) {
+      return null;
+    }
 
-    walletComic = await this.prisma.walletComic.upsert({
-      where: { comicSlug_walletAddress: { walletAddress, comicSlug } },
-      create: { walletAddress, comicSlug, isSubscribed: true },
-      update: { isSubscribed: !walletComic?.isSubscribed },
-    });
-
-    return walletComic;
-  }
-
-  async toggleFavourite(walletAddress: string, comicSlug: string) {
-    let walletComic = await this.prisma.walletComic.findUnique({
-      where: { comicSlug_walletAddress: { walletAddress, comicSlug } },
-    });
+    const [payloadKey, payloadValue] = Object.entries(payload).at(0);
+    const updatePayload =
+      typeof payloadValue === 'boolean'
+        ? {
+          [payloadKey]: !walletComic[payloadKey],
+          }
+        : payload;
 
     walletComic = await this.prisma.walletComic.upsert({
       where: { comicSlug_walletAddress: { walletAddress, comicSlug } },
-      create: { walletAddress, comicSlug, isFavourite: true },
-      update: { isFavourite: !walletComic?.isFavourite },
+      create: { walletAddress, comicSlug, ...payload },
+      update: updatePayload,
     });
```

Comment:
`payload: Record<string, boolean | Date>` is not type safe. Someone can input `isFavorite` instead of `isFavourite` and there would be no lint error due to the general `string` type.
We can fix this type to look something like `Record<keyof WalletComic, boolean | Date` I guess, or you maybe `Partial<WalletComic>`

On the other side, `toggleAction` is not a suitable name for this fuction, it's more like a general `update` function.

Lastly, I have a feeling like we don't want this kind of abstraction.
Do you mind if we switch back to `toggleFavourite`, `toggleSubscribe`, and add `updateDate` or something like that?

Reply:
Agree with adding keyof instead. 

I can switch back, but we are ruing DRY principle, which I don't like, but however you guys want.